### PR TITLE
[Merged by Bors] - feat: port data.mllist

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ListM.lean [hunk 1]
```diff
@@ -16,238 +16,241 @@ Lazy lists with "lazyness" controlled by an arbitrary monad.
...
 -/
 
-
 universe u v
 
 /-- A monadic lazy list, controlled by an arbitrary monad. -/
 unsafe inductive ListM (m : Type u → Type u) (α : Type u) : Type u
-  | nil : ListM m α
-  | cons : m (Option α × ListM m α) → ListM m α
+| nil : ListM m α
```
Comment (jcommelin) [inline Mathlib/Data/ListM.lean:31]: I think mathport got it right, by indenting the constructors, didn't it?

Comment (semorrison) [reply Mathlib/Data/ListM.lean]: I guess so. Weird that we indent after `inductive`, but not after `match`, but okay! Fixed now.

**Review by jcommelin — COMMENTED** [review]
This diff https://github.com/leanprover-community/mathlib4/pull/2945/files/65652264eccd6c45868e0a8bbff58b11bf768c54..684a426f9050fa43f2e07cb6d856d5ddb61ed81f still shows a bunch of places where the indenting differs from mathport output. My impression is that mathport always puts `|` in the right place. (I will not make claims about what happens on other lines, because mathport can by quite trigger-happy with inserting linebreaks and indents in general.)

Otherwise, LGTM

bors d+

File: Mathlib/Data/ListM.lean [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+variable [Monad m]
+
+/-- Repeatedly apply a function `f : α → m (α × List β)` to an initial `a : α`,
+accumulating the elements of the resulting `list β` as a single monadic lazy list.
+
+(This variant allows starting with a specified `List β` of elements, as well. )-/
+unsafe def fixl_with [Alternative m] (f : α → m (α × List β)) : α → List β → ListM m β
+| s, b :: rest => cons <| pure (some b, fixl_with f s rest)
+| s, [] =>
```
Comment (eric-wieser) [inline Mathlib/Data/ListM.lean:52]: Similar here, mathport wants you to indent it seems

File: Mathlib/Data/ListM.lean [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+As we're unsafe anyway, we don't bother with proofs about these constructions.
+-/
+
+universe u v
+
+/-- A monadic lazy list, controlled by an arbitrary monad. -/
+unsafe inductive ListM (m : Type u → Type u) (α : Type u) : Type u
+  | nil : ListM m α
+  | cons : m (Option α × ListM m α) → ListM m α
+#align tactic.mllist ListM
```
Comment (eric-wieser) [inline Mathlib/Data/ListM.lean:33]: Did we drop the extra `l` (lazy) deliberately?

File: Mathlib/Data/ListM.lean [hunk 4]
```diff
@@ -33,221 +36,220 @@ namespace ListM
...
+(This variant allows starting with a specified `List β` of elements, as well. )-/
 unsafe def fixl_with [Alternative m] (f : α → m (α × List β)) : α → List β → ListM m β
-  | s, b :: rest => cons <| pure (some b, fixl_with f s rest)
-  | s, [] =>
-    cons <|
-      (do
-          let (s', l) ← f s
-          match l with
-            | b :: rest => pure (some b, fixl_with f s' rest)
-            | [] => pure (none, fixl_with f s' [])) <|>
```
Comment (eric-wieser) [inline Mathlib/Data/ListM.lean]: It looks like mathport also wants yoy to indent the `match`?

File: Mathlib/Data/ListM.lean (hunk 3 above)
Comment (semorrison) [reply Mathlib/Data/ListM.lean:33]: Arguably whether it is lazy depends on which monad you put here.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**